src/rust/lib/graph-editor/src/component/visualization/fullscreen.rs

```diff
@@ -0,0 +1,338 @@
+use crate::prelude::*;
+
+use crate::component::visualization::traits::{HasSymbols, HasFullscreenDecoration};
+use crate::component::visualization::traits::Resizable;
+
+use ensogl::animation::physics::inertia::DynSimulator;
+use ensogl::control::callback;
+use ensogl::display::Scene;
+use ensogl::display::traits::*;
+use ensogl::display;
+use ensogl::frp;
+use ensogl::gui::component::animation;
+use ensogl::system::web;
+
+
+
+// ====================
+// === Helper Types ===
+// ====================
+
+pub trait Fullscreenable = display::Object+Resizable+HasSymbols+CloneRef+HasFullscreenDecoration
+                           +'static;
+
+
+
+// =========================
+// === Animation Helpers ===
+// =========================
+
+/// Weak ref to the inner `State` of the `StateModel`.
+type WeakState<T> = Weak<RefCell<StateModel<T>>>;
+
+/// Initial and target state of the animated component. This is used in the animation to interpolate between
+/// the initial and the target state.
+#[derive(Debug,Clone)]
+struct AnimationTargetState<T> {
+    /// Animated UI component.
+    target      : T,
+    /// Position at the start of the animation.
+    source_pos  : Vector3<f32>,
+    /// Size at the start of the animation.
+    source_size : Vector3<f32>,
+    /// Position at the end of the animation.
+    target_pos  : Vector3<f32>,
+    /// Size at the end of the animation.
+    target_size : Vector3<f32>,
+}
+
+impl<T:Fullscreenable> AnimationTargetState<T> {
+    /// Set the interpolated values on the target.
+    fn apply_interpolation(&self, value:f32) {
+        let target      = &self.target;
+        let source_pos  = self.source_pos;
+        let source_size = self.source_size;
+        let target_pos  = self.target_pos;
+        let target_size = self.target_size;
+        let pos         = source_pos  * (1.0 - value) + target_pos * value;
+        let size        = source_size * (1.0 - value) + target_size * value;
+        target.set_position(pos);
+        target.set_size(size);
+    }
+}
+
+/// Helper function that contains the logic of the animation.
+fn transition_animation_fn<T:Fullscreenable>(state: WeakState<T>, value:f32) {
+    if let Some(state) = state.upgrade() {
+        // Regular animation step
+        // We animate only if the correct state is given.
+        match state.borrow().deref(){
+            StateModel::TransitioningFromFullscreen {animation_data, .. }
+            | StateModel::TransitioningToFullscreen { animation_data,.. } => {
+                animation_data.apply_interpolation(value);
+            },
+            _  => ()
+        }
+        // Check for end of animation and update the state.
+        if value >= 1.0 {
+           state.borrow_mut().animation_end_transition();
+        }
+    }
+}
+
+// ==================
+// === StateModel ===
+// ==================
+
+/// Represents the internal state of the FullscreenState. This is used to ensure animations always
+/// finish without interruption.
```

Comment:
This is bad. We do want to interrupt animations - look at the cursor press animation. If you press and hold it gets slowly smaller. But if you release the animation starts in the middle of previous naimation.